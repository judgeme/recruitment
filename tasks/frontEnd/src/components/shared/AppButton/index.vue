<script setup lang="ts">
import { defineProps, withDefaults, defineOptions, computed, toRefs } from 'vue';
import type { ButtonHTMLAttributes } from 'vue';
import type { RouteLocationRaw } from 'vue-router';

import AppLink from '@/components/shared/AppLink/index.vue';

// extends ButtonHTMLAttributes to have access to native button props
interface Props extends /* @vue-ignore */ ButtonHTMLAttributes {
  variant?: 'primary' | 'secondary' | 'danger' | 'outline';
  loading?: boolean;
  fullWidth?: boolean;
  disabled?: boolean;
  as?: 'button' | 'a';
  to?: RouteLocationRaw;
}

// set display name in dev-tools
defineOptions({
  name: 'AppButton'
});

const props = withDefaults(defineProps<Props>(), {
  variant: 'primary',
  loading: false,
  disabled: false,
  as: 'button',
  to: ''
});

const { disabled, variant, fullWidth, loading } = toRefs(props);

const buttonClass = `
  py-3 px-6 md:px-9 border-2 border-transparent 
  appearance-none font-inherit inline-block
  align-middle outline-0 text-tiny 
  tracking-compact font-bold uppercase 
  transition-colors text-center
`;

const computedClass = computed(() => {
  return [
    buttonClass,
    {
      'cursor-pointer': !disabled.value,
      'bg-secondary hover:bg-secondary/80 text-primary': variant.value === 'secondary',
      'bg-danger hover:bg-danger/80 text-white': variant.value === 'danger',
      'bg-transparent text-black hover:text-black-80': variant.value === 'outline',
      'bg-primary hover:bg-black-80 text-white': variant.value === 'primary',
      'w-full': fullWidth.value,
      'bg-primary/70 hover:bg-primary/70': loading.value,
      'cursor-not-allowed opacity-70': disabled.value
    }
  ];
});
</script>
<template>
  <app-link v-if="as === 'a'" :to="to" :class="computedClass">
    <slot />
  </app-link>
  <button v-else :class="computedClass" :disabled="disabled">
    <slot />
  </button>
</template>
