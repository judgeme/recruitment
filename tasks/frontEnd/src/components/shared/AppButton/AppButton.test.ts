import { mount } from '@vue/test-utils';
import { createRouter, createWebHistory } from 'vue-router';
import { describe, it, expect } from 'vitest';

import AppButton from '@/components/shared/AppButton/index.vue'; 

// Create a mock router
const router = createRouter({
  history: createWebHistory(),
  routes: [{ path: '/', component: { template: '<div>Home</div>' } }],
});

describe('AppButton.vue', () => {
  it('renders button element correctly with default props', () => {
    const wrapper = mount(AppButton);

    // Check if the button element is rendered
    const button = wrapper.find('button');
    expect(button.exists()).toBe(true);

    // Check if default classes are applied
    expect(button.classes()).toContain('py-3');
    expect(button.classes()).toContain('px-6');
    expect(button.classes()).toContain('bg-primary');
    expect(button.classes()).toContain('text-white');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('applies variant classes correctly', async () => {
    const wrapper = mount(AppButton, {
      props: { variant: 'secondary' },
    });

    const button = wrapper.find('button');
    expect(button.classes()).toContain('bg-secondary');
    expect(button.classes()).toContain('text-primary');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders as anchor element with correct props', async () => {
    const wrapper = mount(AppButton, {
      props: { as: 'a', to: '/path' },
      global: {
        plugins: [router]
      }
    });

    const appLink = wrapper.findComponent({ name: 'AppLink' });
    expect(appLink.exists()).toBe(true);
    expect(appLink.props('to')).toBe('/path');

    expect(wrapper.html()).toMatchSnapshot();
  });

});
