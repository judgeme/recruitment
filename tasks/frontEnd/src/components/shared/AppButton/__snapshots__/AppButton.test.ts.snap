// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppButton.vue > applies variant classes correctly 1`] = `
"<button class="py-3 px-6 md:px-9 border-2 border-transparent 
  appearance-none font-inherit inline-block
  align-middle outline-0 text-tiny 
  tracking-compact font-bold uppercase 
  transition-colors text-center cursor-pointer bg-secondary hover:bg-secondary/80 text-primary"></button>"
`;

exports[`AppButton.vue > renders as anchor element with correct props 1`] = `
"<a class="py-3 px-6 md:px-9 border-2 border-transparent 
  appearance-none font-inherit inline-block
  align-middle outline-0 text-tiny 
  tracking-compact font-bold uppercase 
  transition-colors text-center cursor-pointer bg-primary hover:bg-black-80 text-white" href="/path"></a>"
`;

exports[`AppButton.vue > renders button element correctly with default props 1`] = `
"<button class="py-3 px-6 md:px-9 border-2 border-transparent 
  appearance-none font-inherit inline-block
  align-middle outline-0 text-tiny 
  tracking-compact font-bold uppercase 
  transition-colors text-center cursor-pointer bg-primary hover:bg-black-80 text-white"></button>"
`;
